packages/components/src/Modal.ts

```diff
@@ -0,0 +1,245 @@
+  const previouslyFocusedElement = useRef<Element | null>(null);
+  useEffect(() => {
+    previouslyFocusedElement.current = document.activeElement;
+
+    // When the modal is closed (unmounted), return focus to the previously focused element.
+    return () => {
+      if (previouslyFocusedElement.current instanceof HTMLElement) {
+        previouslyFocusedElement.current.focus();
+      }
+    };
+  }, []);
+
+  return h(
+    RModal,
+    {
+      aria: {
+        labelledby: hasTitle ? titleId : undefined,
+        modal: true,
+      },
+      // Adding aria-hidden to the app container is unnecessary with aria-modal set on the modal.
+      // See https://github.com/DataBiosphere/terra-ui/pull/2541
+      ariaHideApp: false,
+      contentRef: (node: HTMLDivElement) => {
+        modalElement.current = node;
+      },
+      isOpen: true,
+      parentSelector: getPopupRoot,
+      // Customize focus handling using onAfterOpen to properly manage focus where
+      // react-modal and react-focus-lock interact.
+      // See https://github.com/DataBiosphere/terra-ui/pull/1938
+      shouldFocusAfterRender: false,
+      shouldReturnFocusAfterClose: false,
+      style: {
+        overlay: modalStyles.overlay,
+        content: {
+          ...modalStyles.modal,
+          width,
+          ...styles?.modal,
+        },
+      },
+      onAfterOpen: () => {
+        // Since this is called right after the Modal is mounted, modalElement.current should always be non-null here.
+        const nodeToFocus = modalElement.current!.contains(document.activeElement)
+          ? document.activeElement
+          : modalElement.current;
+
+        // Add the focus update to the end of the event queue.
+        // Per react-focus-lock: https://github.com/theKashey/react-focus-lock#unmounting-and-focus-management
+        setTimeout(() => {
+          // Store the previously focused element so that focus can be returned to it after the modal is closed.
+          previouslyFocusedElement.current = modalElement.current?.contains(document.activeElement)
+            ? previouslyFocusedElement.current
+            : document.activeElement;
+
+          // modalElement.current may be null here if the modal has been unmounted before this is reached.
+          // This is most likely to occur in tests.
+          const isModalMounted = modalElement.current !== null;
+          if (isModalMounted) {
+            const isFocusWithinModal =
+              !!document.activeElement &&
+              (modalElement.current === document.activeElement ||
+                modalElement.current!.contains(document.activeElement));
+
+            // If focus is already within the modal, leave it alone.
+            // This avoids issues in tests where focus is unexpectedly moved after an interaction.
+            if (!isFocusWithinModal && document.contains(nodeToFocus) && nodeToFocus instanceof HTMLElement) {
```

Comment:
This resolved some issues in tests. In some cases, the test would render a modal and click to open a menu in the modal. Then `onAfterOpen` would move focus, closing the menu. The test would assume the menu was open since it was just clicked and fail.